File: UserNotifications/UserNotificationsObjCSample/UserNotificationsObjCSample/UserNotificationsHandler.m
//
//  UserNotificationsHandler.m
//  UserNotificationsObjCSample
//
//  Created by Nikolay Andonov on 11/9/16.
//  Copyright © 2016 Mentormate. All rights reserved.
//

#import "UserNotificationsHandler.h"
#import <UserNotifications/UserNotifications.h>

static NSString *const RemindLaterActionIdentifier = @"remindLater";
static NSString *const LogoAttachmentIdentifier = @"logo";
static NSString *const NotificationCategoryIdentifier = @"myCategory";
static NSString *const NotificationRequestIdentifier = @"myNotificationRequest";

@interface UserNotificationsHandler() <UNUserNotificationCenterDelegate>

@end

@implementation UserNotificationsHandler

+ (instancetype)sharedHandler {
    
    static UserNotificationsHandler *userNotificationHandler = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        userNotificationHandler = [[UserNotificationsHandler alloc] init];
    });
    
    return userNotificationHandler;
}

- (void)userNotificationsInitialSetup {
    
    [[UNUserNotificationCenter currentNotificationCenter] requestAuthorizationWithOptions: (UNAuthorizationOptionSound | UNAuthorizationOptionAlert) completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if (!granted) {
            NSLog(@"Notification access denied.");
        }
    }];
    
    UNNotificationAction *action = [UNNotificationAction actionWithIdentifier:RemindLaterActionIdentifier title:@"Remind me later" options:0];
    UNNotificationCategory *category = [UNNotificationCategory categoryWithIdentifier:NotificationCategoryIdentifier actions:@[action] intentIdentifiers:@[] options:0];
    
    [[UNUserNotificationCenter currentNotificationCenter] setNotificationCategories:[NSSet setWithObject:category]];
}

- (void)scheduleNotificationAtDate:(NSDate *)date {

    NSCalendar *calendar = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSDateComponents *components = [calendar componentsInTimeZone:[NSTimeZone localTimeZone] fromDate:date];
    NSDateComponents *newComponents = [[NSDateComponents alloc] init];
    newComponents.calendar = calendar;
    newComponents.timeZone = [NSTimeZone localTimeZone];
    newComponents.month = components.month;
    newComponents.day = components.day;
    newComponents.hour = components.hour;
    newComponents.minute = components.minute;
    
    
    UNCalendarNotificationTrigger *trigger = [UNCalendarNotificationTrigger triggerWithDateMatchingComponents:newComponents repeats:NO];
    UNMutableNotificationContent *content = [[UNMutableNotificationContent alloc] init];
    
    content.title = @"Reminder";
    content.body = @"Reminder to do some generic activity";
    content.sound = [UNNotificationSound defaultSound];
    content.categoryIdentifier = NotificationCategoryIdentifier;
    
    NSString *path = [[NSBundle mainBundle] pathForResource:@"logo" ofType: @"png"];
    if (path) {
        
        NSURL *url = [NSURL fileURLWithPath:path];
        NSError *attachmentError = nil;
        UNNotificationAttachment *attachment = [UNNotificationAttachment attachmentWithIdentifier:LogoAttachmentIdentifier URL:url options:nil error:&attachmentError];
        if (attachmentError) {
            NSLog(@"The attachment was not loaded.");
        }
        else {
            content.attachments = @[attachment];
        }
    }
    
    UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:NotificationRequestIdentifier content:content trigger:trigger];
    [UNUserNotificationCenter currentNotificationCenter].delegate = self;
    [[UNUserNotificationCenter currentNotificationCenter] removeAllPendingNotificationRequests];
    [[UNUserNotificationCenter currentNotificationCenter] addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
        if (error) {
            NSLog(@"Error occured: %@", error);
        }
    }];
    
}

#pragma mark - UNUserNotificationCenterDelegate

- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler {
    
    completionHandler();
    if ([response.actionIdentifier isEqualToString: RemindLaterActionIdentifier]) {
        NSDate *newDate = [NSDate dateWithTimeInterval: 5 * 60 sinceDate:[NSDate date]];
        [self scheduleNotificationAtDate:newDate];
    }
}



@end

File: UserNotifications/UserNotificationsSwiftSample/UserNotificationsSwiftSample/UserNotificationsHandler.swift
//
//  UserNotificationsHandler.swift
//  UserNotificationsSwiftSample
//
//  Created by Nikolay Andonov on 11/9/16.
//  Copyright © 2016 Mentormate. All rights reserved.
//

import UIKit
import UserNotifications

private let sharedInstance = UserNotificationsHandler()

let RemindLaterActionIdentifier = "remindLater"
let LogoAttachmentIdentifier = "logo"
let NotificationCategoryIdentifier = "myCategory"
let NotificationRequestIdentifier = "myNotificationRequest"


class UserNotificationsHandler: NSObject, UNUserNotificationCenterDelegate {

    class var sharedHandler: UserNotificationsHandler {
        return sharedInstance
    }
    
    func userNotificationsInitialSetup() {
        
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) {(accepted, error) in
            if !accepted {
                print("Notification access denied.")
            }
        }
        
        let action = UNNotificationAction(identifier: RemindLaterActionIdentifier, title: "Remind me later", options: [])
        let category = UNNotificationCategory(identifier: NotificationCategoryIdentifier, actions: [action], intentIdentifiers: [], options: [])
        UNUserNotificationCenter.current().setNotificationCategories([category])
    }
    
    func scheduleNotification(at date: Date) {
        
        let calendar = Calendar(identifier: .gregorian)
        let components = calendar.dateComponents(in: .current, from: date)
        let newComponents = DateComponents(calendar: calendar, timeZone: .current, month: components.month, day: components.day, hour: components.hour, minute: components.minute)
        
        let trigger = UNCalendarNotificationTrigger(dateMatching: newComponents, repeats: false)
        
        let content = UNMutableNotificationContent()
        content.title = "Reminder"
        content.body = "Reminder to do some generic activity"
        content.sound = UNNotificationSound.default()
        content.categoryIdentifier = NotificationCategoryIdentifier
        
        if let path = Bundle.main.path(forResource: "logo", ofType: "png") {
            let url = URL(fileURLWithPath: path)
            
            do {
                let attachment = try UNNotificationAttachment(identifier: LogoAttachmentIdentifier, url: url, options: nil)
                content.attachments = [attachment]
            } catch {
                print("The attachment was not loaded.")
            }
        }
        
        let request = UNNotificationRequest(identifier: NotificationRequestIdentifier, content: content, trigger: trigger)
        
        UNUserNotificationCenter.current().delegate = self
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
        UNUserNotificationCenter.current().add(request) {(error) in
            if let error = error {
                print("Error occured: \(error)")
            }
        }
    }

    //MARK: - UNUserNotificationCenterDelegate
    
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        
        completionHandler()
        if response.actionIdentifier == RemindLaterActionIdentifier {
            let newDate = Date(timeInterval: 60 * 5, since: Date())
            scheduleNotification(at: newDate)
        }
    }
}
